### PR TITLE
Added services in hostvars

### LINKED ISSUE #47918: services-addition in hostvars in netbox dynamic inventory
##### SUMMARY
This is an extension to https://github.com/ansible/ansible/issues/47391
Adding services defined in Netbox to host vars would assist those intending to use Netbox to configure monitoring systems.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
netbox-services in hostvars in netbox dynamic inventory

##### ADDITIONAL INFORMATION
Feature could be used to configure monitoring systems by addition of services defined in Netbox in Ansible hostvars. Netbox allows arbitrary structured data to be assigned to servers via custom fields, tags, configuration contexts and services. Along with https://github.com/ansible/ansible/pull/47394 this would complete the set of these Netbox features support by the dynamic inventory script.

```yaml
ansible-inventory -vvv -i ./netbox_inventory.yml --host someserver
ansible-inventory 2.7.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/monitoring/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-inventory
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
Using /etc/ansible/ansible.cfg as config file
/home/monitoring/ansible/netbox_inventory.yml did not meet host_list requirements, check plugin documentation if this is unexpected
/home/monitoring/ansible/netbox_inventory.yml did not meet virtualbox requirements, check plugin documentation if this is unexpected
Fetching: https://netboxdev.hyperslice.net/api/dcim/sites/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/regions/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/tenancy/tenants/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/racks/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-roles/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-types/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/manufacturers/?limit=0
Fetching: https://netboxdev.hyperslice.net//api/dcim/devices/?limit=0&name=someserver
Fetching: https://netboxdev.hyperslice.net//api/virtualization/virtual-machines/?limit=0&name=someserver
Parsed /home/monitoring/ansible/netbox_inventory.yml inventory source with netbox plugin
{
    "config_context": [
        {}
    ],
    "device_roles": [
        "Dummy server"
    ],
    "device_types": [
        "Dummy"
    ],
    "manufacturers": [
        "Dummy"
    ],
    "services": [
        [
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 28,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:01.696607Z",
                "name": "SSH",
                "port": 22,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 29,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:18.855776Z",
                "name": "awebsite.org",
                "port": 443,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 30,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:32.669569Z",
                "name": "anothersite.net",
                "port": 443,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 31,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:54.599007Z",
                "name": "arbitaryservice",
                "port": 5544,
                "protocol": {
                    "label": "UDP",
                    "value": 17
                },
                "virtual_machine": null
            }
        ]
    ],
    "sites": [
        "Dummy site"
    ]
}
```

### PR DESCRIPTION
**Pull request by ollybee**
##### SUMMARY
services addition in hostvars in netbox dynamic inventory
Fixed #47918

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

netbox-services in hostvars in netbox dynamic inventory

##### ANSIBLE VERSION

```paste below
ansible 2.7.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/oliver/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
```

##### ADDITIONAL INFORMATION
Before change
```
ansible-inventory 2.7.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/monitoring/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-inventory
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
Using /etc/ansible/ansible.cfg as config file
/home/monitoring/ansible/netbox_inventory.yml did not meet host_list requirements, check plugin documentation if this is unexpected
/home/monitoring/ansible/netbox_inventory.yml did not meet virtualbox requirements, check plugin documentation if this is unexpected
Fetching: https://netboxdev.hyperslice.net/api/dcim/sites/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/regions/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/tenancy/tenants/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/racks/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-roles/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-types/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/manufacturers/?limit=0
Fetching: https://netboxdev.hyperslice.net//api/dcim/devices/?limit=0&name=someserver
Fetching: https://netboxdev.hyperslice.net//api/virtualization/virtual-machines/?limit=0&name=someserver
Parsed /home/monitoring/ansible/netbox_inventory.yml inventory source with netbox plugin
{
    "config_context": [
        {}
    ],
    "device_roles": [
        "Dummy server"
    ],
    "device_types": [
        "Dummy"
    ],
    "manufacturers": [
        "Dummy"
    ],
    "sites": [
        "Dummy site"
    ]
}
```
After Change
```
ansible-inventory -vvv -i ./netbox_inventory.yml --host someserver
ansible-inventory 2.7.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/monitoring/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-inventory
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
Using /etc/ansible/ansible.cfg as config file
/home/monitoring/ansible/netbox_inventory.yml did not meet host_list requirements, check plugin documentation if this is unexpected
/home/monitoring/ansible/netbox_inventory.yml did not meet virtualbox requirements, check plugin documentation if this is unexpected
Fetching: https://netboxdev.hyperslice.net/api/dcim/sites/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/regions/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/tenancy/tenants/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/racks/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-roles/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/device-types/?limit=0
Fetching: https://netboxdev.hyperslice.net/api/dcim/manufacturers/?limit=0
Fetching: https://netboxdev.hyperslice.net//api/dcim/devices/?limit=0&name=someserver
Fetching: https://netboxdev.hyperslice.net//api/virtualization/virtual-machines/?limit=0&name=someserver
Parsed /home/monitoring/ansible/netbox_inventory.yml inventory source with netbox plugin
{
    "config_context": [
        {}
    ],
    "device_roles": [
        "Dummy server"
    ],
    "device_types": [
        "Dummy"
    ],
    "manufacturers": [
        "Dummy"
    ],
    "services": [
        [
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev.hyperslice.net/api/dcim/devices/13084/"
                },
                "id": 28,
                "ipaddresses": [
                    {
                        "address": "10.2.1.1/24",
                        "family": 4,
                        "id": 145531,
                        "url": "https://netboxdev.hyperslice.net/api/ipam/ip-addresses/145531/"
                    }
                ],
                "last_updated": "2018-11-01T12:13:39.326592Z",
                "name": "SSH",
                "port": 22,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 29,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:18.855776Z",
                "name": "awebsite.org",
                "port": 443,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 30,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:32.669569Z",
                "name": "anothersite.net",
                "port": 443,
                "protocol": {
                    "label": "TCP",
                    "value": 6
                },
                "virtual_machine": null
            },
            {
                "created": "2018-11-01",
                "custom_fields": {
                    "key_text": null
                },
                "description": "",
                "device": {
                    "display_name": "someserver",
                    "id": 13084,
                    "name": "someserver",
                    "url": "https://netboxdev/api/dcim/devices/13084/"
                },
                "id": 31,
                "ipaddresses": [],
                "last_updated": "2018-11-01T11:59:54.599007Z",
                "name": "arbitaryservice",
                "port": 5544,
                "protocol": {
                    "label": "UDP",
                    "value": 17
                },
                "virtual_machine": null
            }
        ]
    ],
    "sites": [
        "Dummy site"
    ]
}
```

### CODE REVIEW

**Comment by FragmentedPacket** [comment]
@ollybee I would most likely remove the `[]` around the `device_lookup["results"]` as it's already a list so we wouldn't have to hit a 0 index to get the necessary information.

Also, you may want to sync up your file with what is in devel for the `def refresh_url` as it contains some previous bug fixes and we don't regress.

Other than those, it worked as intended:
```
            "test51": {
                "config_context": [
                    {
                        "test": "test1"
                    }
                ], 
                "device_roles": [
                    "Console Server"
                ], 
                "device_types": [
                    "test"
                ], 
                "manufacturers": [
                    "Cisco"
                ], 
                "services": [
                    {
                        "created": "2018-11-02", 
                        "custom_fields": {}, 
                        "description": "", 
                        "device": {
                            "display_name": "test51", 
                            "id": 74, 
                            "name": "test51", 
                            "url": "http://netbox-demo.org/api/dcim/devices/74/"
                        }, 
                        "id": 1, 
                        "ipaddresses": [], 
                        "last_updated": "2018-11-02T02:00:43.567445Z", 
                        "name": "Jalepenos", 
                        "port": 144, 
                        "protocol": {
                            "label": "TCP", 
                            "value": 6
                        }, 
                        "virtual_machine": null
                    }, 
                    {
                        "created": "2018-11-02", 
                        "custom_fields": {}, 
                        "description": "", 
                        "device": {
                            "display_name": "test51", 
                            "id": 74, 
                            "name": "test51", 
                            "url": "http://netbox-demo.org/api/dcim/devices/74/"
                        }, 
                        "id": 2, 
                        "ipaddresses": [], 
                        "last_updated": "2018-11-02T02:01:45.851961Z", 
                        "name": "Pineapples", 
                        "port": 1412, 
                        "protocol": {
                            "label": "UDP", 
                            "value": 17
                        }, 
                        "virtual_machine": null
                    }
                ], 
                "sites": [
                    "test"
                ]
            }
```

**Comment by FragmentedPacket** [comment]
LGTM

File: lib/ansible/plugins/inventory/netbox.py [hunk 1]
```diff
@@ -198,6 +198,8 @@ def group_extractors(self):
             "device_roles": self.extract_device_role,
             "platforms": self.extract_platform,
             "device_types": self.extract_device_type,
+            "config_context": self.extract_config_context,
```
Comment (nikkytub) [inline lib/ansible/plugins/inventory/netbox.py]: This is the change proposed by me at https://github.com/ansible/ansible/pull/47394 which hasn't been merged yet. Why have you included it here?

File: lib/ansible/plugins/inventory/netbox.py [hunk 2]
```diff
@@ -246,6 +248,22 @@ def extract_device_role(self, host):
         except Exception:
             return
 
+    def extract_config_context(self, host):
```
Comment (nikkytub) [inline lib/ansible/plugins/inventory/netbox.py]: Addition of my changes will be seen as yours which is plagiarism

Comment (nikkytub) [reply lib/ansible/plugins/inventory/netbox.py]: I would suggest you to remove the config-context part and keep your services feature here

Comment (ollybee) [reply lib/ansible/plugins/inventory/netbox.py]: Thank you. I've removed your code to leave only the services feature.

**Comment by FragmentedPacket** [comment]
LGTM

**Comment by gundalow** [comment]
@ollybee Can you please rebase https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html

**Comment by FragmentedPacket** [comment]
@ollybee Do you mind rebasing and then we can get this merged? 2.8 RC is right around the corner.

**Comment by jhanssenfavaro** [comment]
Hi Folks, I am also waiting for this commit. Its a huge improvement in the netbox plugin.. anyone aware if is it already available ? Best Regards!!!

**Comment by Akasurde** [comment]
superseded by https://github.com/netbox-community/ansible_modules/pull/202

@ollybee Thanks for the contribution.

**Pull request closed without merging**